### PR TITLE
Adopt v3 APIs related with services

### PR DESCRIPTION
**Pull request by theghost5800**
Adoption is related with get and list methods of service instance and
full usage of v3 service broker api.

JIRA:LMCROSSITXSADEPLOY-2251

### CODE REVIEW

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 1]
```diff
@@ -1719,68 +1693,70 @@ private CloudServiceInstance findServiceInstanceByName(String name) {
...
-                   .map(tuple -> ImmutableRawCloudServiceInstance.builder()
-                                                                 .resource(tuple.getT1())
-                                                                 .serviceResource(tuple.getT2())
-                                                                 .servicePlanResource(tuple.getT3())
-                                                                 .build());
+        String servicePlanGuid = serviceInstanceResource.getRelationships()
+                                                        .getServicePlan()
+                                                        .getData()
+                                                        .getId();
+        Mono<? extends ServicePlan> servicePlanMono = getServicePlanResource(servicePlanGuid);
```
Comment (IvanBorislavovDimitrov) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1744]: servicePlanMono variable is not really needed

File: src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java [hunk 2]
```diff
@@ -88,7 +88,19 @@ void bindServiceInstance(String applicationName, String serviceInstanceName, Map
...
+     *
+     * @param applicationName application name
+     * @param staging staging info /** Create application.
+     *
+     * @param applicationName application name
+     * @param staging staging info
+     * @param disk disk quota to use in MB
+     * @param memory memory to use in MB
+     * @ping staging, Integer disk, Integer memory, Set<CloudRouteSummary> routes, DockerInfo dockerInfo); void createApplication(String
+     *       applicationName, Staging staging, Integer disk, Integer memory, Set<CloudRouteSummary> routes);
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java:101]: Isn't this method commented out?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java [hunk 3]
```diff
@@ -11,21 +15,26 @@
...
 
     @Value.Parameter
-    public abstract Resource<ServiceBrokerEntity> getResource();
+    public abstract ServiceBroker getResource();
 
     @Override
     public CloudServiceBroker derive() {
-        Resource<ServiceBrokerEntity> resource = getResource();
-        ServiceBrokerEntity entity = resource.getEntity();
+        ServiceBroker resource = getResource();
```
[the remark this answers is not in the document]
Comment (radito3) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java:22]: +1

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java [hunk 4]
```diff
@@ -11,21 +15,26 @@
...
-                                          .spaceGuid(entity.getSpaceId())
+                                          .name(resource.getName())
+                                          .url(resource.getUrl())
+                                          .spaceGuid(spaceGuid)
                                           .build();
     }
 
+    private String getSpaceGuid(ServiceBroker resource) {
+        return Optional.ofNullable(resource.getRelationships())
+                       .map(ServiceBrokerRelationships::getSpace)
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java:34]: Is there a possibility that the `relationships` json will be present but the `space` won't?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java [hunk 5]
```diff
@@ -14,28 +13,31 @@
...
-                                            .extra(entity.getExtra())
-                                            .docUrl(entity.getDocumentationUrl())
-                                            .infoUrl(entity.getInfoUrl())
-                                            .version(entity.getVersion())
-                                            .provider(entity.getProvider())
-                                            .brokerName(entity.getServiceBrokerName())
-                                            .uniqueId(entity.getUniqueId())
-                                            .url(entity.getUrl())
+                                            .name(resource.getName())
+                                            .isActive(resource.getAvailable())
```
[the remark this answers is not in the document]
Comment (radito3) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java:26]: +1

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java [hunk 6]
```diff
@@ -14,28 +13,31 @@
...
+                                            .isBindable(resource.getBrokerCatalog()
+                                                                .getFeatures()
+                                                                .getBindable())
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java:29]: Could these methods return a null object?

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/ServiceInstanceTypeDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.cloudfoundry.client.v3.serviceinstances.ServiceInstanceType;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+
+/*
+ * TODO: This custom deserializer will be needed during switch applying of new version, after that this logic can be removed.
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/adapters/ServiceInstanceTypeDeserializer.java:13]: This sentence is a bit confusing. Maybe reword it to
```
This custom deserializer is needed during rollout of the upcoming cf-java-client version 2.12.0.
It can be deleted with version 2.13.0
```

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 8]
```diff
@@ -477,13 +480,23 @@ public void createServiceInstance(CloudServiceInstance serviceInstance) {
...
         Assert.notNull(serviceBroker, "Service broker must not be null.");
 
-        delegate.serviceBrokers()
+        ServiceBrokerRelationships serviceBrokerRelationship = Optional.ofNullable(serviceBroker.getSpaceGuid())
+                                                                       .map(UUID::fromString)
+                                                                       .map(this::buildToOneRelationship)
+                                                                       .map(spaceRelationship -> ServiceBrokerRelationships.builder()
+                                                                                                                           .space(spaceRelationship)
+                                                                                                                           .build())
+                                                                       .orElse(null);
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:489]: nice

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 9]
```diff
@@ -1082,35 +1064,25 @@ public void updateServiceTags(String serviceName, List<String> tags) {
 
     @Override
     public List<CloudServiceInstance> getServiceInstancesByMetadataLabelSelector(String labelSelector) {
-        Map<String, Metadata> serviceInstancesMetadata = getServiceInstancesMetadataByLabelSelector(labelSelector);
-        List<CloudServiceInstance> serviceInstances = fetchListWithAuxiliaryContent(() -> getServiceInstanceResourcesByNamesInBatches(serviceInstancesMetadata.keySet()),
-                                                                                    this::zipWithAuxiliaryServiceInstanceContent);
-        return getServiceInstancesWithMetadata(serviceInstances, serviceInstancesMetadata);
-    }
+        IntFunction<ListServiceInstancesRequest> listServiceInstancesPageRequestSupplier = page -> ListServiceInstancesRequest.builder()
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1067]: This variable name seems a bit too long. `pageSupplier` or `pageRequestSupplier` will be sufficient in my opinion

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 10]
```diff
@@ -1719,68 +1693,70 @@ private CloudServiceInstance findServiceInstanceByName(String name) {
         return fetchWithAuxiliaryContent(() -> getServiceInstanceResourceByName(name), this::zipWithAuxiliaryServiceInstanceContent);
     }
 
-    private Flux<? extends Resource<UnionServiceInstanceEntity>> getServiceInstanceResources() {
-        IntFunction<ListSpaceServiceInstancesRequest> pageRequestSupplier = page -> ListSpaceServiceInstancesRequest.builder()
-                                                                                                                    .returnUserProvidedServiceInstances(true)
-                                                                                                                    .spaceId(getTargetSpaceGuid().toString())
-                                                                                                                    .page(page)
-                                                                                                                    .build();
+    private Flux<? extends ServiceInstanceResource> getServiceInstanceResources() {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:1696]: Isn't `Flux<ServiceInstanceResource>` enough? I think the `? extends` is only needed with `? extends ServiceInstance`

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 11]
```diff
@@ -2516,45 +2509,39 @@ private CloudServiceBroker findServiceBrokerByName(String name) {
...
-                    if (plan.getName()
-                            .equals(planName)) {
-                        return plan;
-                    }
-                }
-            }
-        }
-        throw new CloudOperationException(HttpStatus.NOT_FOUND, "Not Found", "Service plan " + service.getPlan() + " not found.");
+        List<CloudServiceOffering> offerings = service.getBroker() == null ? findServiceOfferingsByLabel(service.getLabel())
+            : findServiceOfferingsByLabelAndBrokerName(service.getLabel(), service.getBroker());
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:2534]: This is hard to read. Split into condition and assignment

File: src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java [hunk 12]
```diff
@@ -2516,45 +2509,39 @@ private CloudServiceBroker findServiceBrokerByName(String name) {
...
+
+        return offerings.stream()
+                        .map(CloudServiceOffering::getServicePlans)
+                        .flatMap(List::stream)
+                        .filter(plan -> plan.getName()
+                                            .equals(planName))
+                        .findFirst()
+                        .orElseThrow(() -> new CloudOperationException(HttpStatus.NOT_FOUND,
+                                                                       "Not Found",
+                                                                       "Service plan " + service.getPlan() + " not found."));
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/rest/CloudControllerRestClientImpl.java:2544]: This is fancy and all but maybe a regular nested loop would be more readable.
```java
for (var offering : offerings) {
   for (var plan : offering.getPlans()) {
      if (planName.equals(plan.getName()) {
         return plan;
      }
  }
}
throw exception()
```

**Review by radito3 — CHANGES_REQUESTED** [review]

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java [hunk 13]
```diff
@@ -14,28 +13,31 @@
...
+                                            .description(resource.getDescription())
+                                            .isShareable(resource.getShareable())
+                                            .extra(resource.getBrokerCatalog()
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java:32]: Yes, it was described here - http://v3-apidocs.cloudfoundry.org/version/3.106.0/index.html#service-bindings-in-v3 as replacement of `extra` field.

File: src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java (hunk 2 above)
Comment (theghost5800) [reply src/main/java/com/sap/cloudfoundry/client/facade/CloudControllerClient.java:101]: nice catch, this mistake was done during merge conflict resolving

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java (hunk 4 above)
Comment (theghost5800) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceBroker.java:34]: Yes, in case service broker is registered as global, then space is omitted - http://v3-apidocs.cloudfoundry.org/version/3.106.0/index.html#the-service-broker-object

File: src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java (hunk 6 above)
Comment (theghost5800) [reply src/main/java/com/sap/cloudfoundry/client/facade/adapters/RawCloudServiceOffering.java:29]: I didn't find null field for broker catalog when execute `GET v3/service_offerings` . In documentation I don't find anything described that it is possible to be null.

**Review by radito3 — APPROVED** [review]

**Review by radito3 — APPROVED** [review]

**Review by radito3 — APPROVED** [review]

**Review by abdermendz — APPROVED** [review]
